### PR TITLE
test/recipes/01-test_symbol_presence.t: check for duplicate symbols in static libs

### PR DESCRIPTION
**Pull request by levitte**
This checks that all symbols are unique across all public static libraries.
This includes a bit of refacftoring to avoid repeating code too much.

-----

This is an alternative for #20308

### CODE REVIEW

**Comment by levitte** [comment]
Ping!  I personally favor this over #20308

File: test/recipes/01-test_symbol_presence.t [hunk 1]
```diff
@@ -23,54 +24,97 @@ use platform;
...
+foreach (sort keys %stlibname) {
+    my $stlib_cmd = "nm -Pg $stlibpath{$_} 2> /dev/null";
+    my $shlib_cmd = "nm -DPg $shlibpath{$_} 2> /dev/null";
+    my @stlib_lines;
+    my @shlib_lines;
+    *OSTDERR = *STDERR;
+    *OSTDOUT = *STDOUT;
+    open STDERR, ">", devnull();
+    open STDOUT, ">", devnull();
+    @stlib_lines = map { s|\R$||; $_ } `$stlib_cmd`; $? = 1;
```
Comment (mattcaswell) [inline test/recipes/01-test_symbol_presence.t:75]: I don't understand the trailing `$? = 1` here. Can you explain it?

File: test/recipes/01-test_symbol_presence.t [hunk 2]
```diff
@@ -23,54 +24,97 @@ use platform;
...
+    if ($? != 0) {
+        note "running '$stlib_cmd' => $?";
+        @stlib_lines = ();
+    }
+    unless (disabled('shared')) {
+        @shlib_lines = map { s|\R$||; $_ } `$shlib_cmd`; $? = 1;
```
Comment (mattcaswell) [inline test/recipes/01-test_symbol_presence.t:81]: And here

File: test/recipes/01-test_symbol_presence.t [hunk 3]
```diff
@@ -23,54 +24,97 @@ use platform;
...
+    my @def_lines;
+    unless (disabled('shared')) {
         indir $bldtop => sub {
             my $mkdefpath = srctop_file("util", "mkdef.pl");
-            my $libnumpath = srctop_file("util", "lib$libname.num");
-            @def_lines = map { s|\R$||; $_ } `$^X $mkdefpath --ordinals $libnumpath --name $libname --OS linux 2> /dev/null`;
-            ok($? == 0, "running 'cd $bldtop; $^X $mkdefpath --ordinals $libnumpath --name $libname --OS linux' => $?");
+            my $def_path = srctop_file("util", "lib$_.num");
+            my $def_cmd = "$^X $mkdefpath --ordinals $def_path --name $_ --OS linux 2> /dev/null";
+            @def_lines = map { s|\R$||; $_ } `$def_cmd`; $? = 1;
```
Comment (mattcaswell) [inline test/recipes/01-test_symbol_presence.t:99]: And here

File: test/recipes/01-test_symbol_presence.t (hunk 1 above)
Comment (levitte) [reply test/recipes/01-test_symbol_presence.t:75]: Yes I can!  It's a remnant from an incantation that never made it into this PR.  Thanks for spotting it, I'm now removing it (at every spot)

File: test/recipes/01-test_symbol_presence.t [hunk 4]
```diff
@@ -23,54 +24,97 @@ use platform;
...
-        close STDOUT;
-        *STDERR = *OSTDERR;
-        *STDOUT = *OSTDOUT;
-        skip "Can't run 'nm -DPg $shlibpath' => $?...  ignoring", 2
-            unless $? == 0;
-
-        my $bldtop = bldtop_dir();
-        my @def_lines;
+    "ordinals (util/*.num in the source tree), and that our static libraries\n",
+    "don't shared symbols, something that should be good enough a check for\n",
```
[the remark this answers is not in the document]
Comment (levitte) [reply test/recipes/01-test_symbol_presence.t]: Fixed

**Review by mattcaswell — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged.

**Pull request closed without merging**